Add support for SENSOR_ATTR_OFFSET, multiple instances, and update Kconfig for TMP117

puranjaymohan opened the pull request:
Added support for SENSOR_ATTR_OFFSET and sample code for the [TI TMP117](https://www.ti.com/lit/ds/symlink/tmp117.pdf?ts=1607855698952). This PR also adds support for multiple instances, updates Kconfig to include TMP117.

This closes #30663

mniestroj commented on drivers/sensor/tmp117/tmp117.c:
```diff
@@ -0,0 +1,189 @@
+	.i2c_addr = DT_INST_REG_ADDR(0),
+};
+
+DEVICE_AND_API_INIT(tmp117, DT_INST_LABEL(0), tmp117_init,
```
Use `DEVICE_DT_INST_DEFINE` instead. See #30766.

mniestroj commented on drivers/sensor/tmp117/tmp117.c:
```diff
@@ -0,0 +1,189 @@
+	int rc;
+
+	/* Bind to the I2C bus that the sensor is connected */
+	drv_data->i2c = device_get_binding(DT_INST_BUS_LABEL(0));
```
Would be nice to save `DT_INST_BUS_LABEL(0)` into `struct tmp117_dev_config tmp117_config `, same as `i2c_addr` is.

mniestroj commented on drivers/sensor/tmp117/tmp117.c:
```diff
@@ -0,0 +1,189 @@
+
+	if (tmp117_reg_read(dev, TMP117_REG_DEVICE_ID, &value) != 0) {
+		LOG_ERR("%s: Failed to get Device ID register!",
+			DT_INST_LABEL(0));
```
Please use `dev->name` instead of `DT_INST_LABEL(0)`.

mniestroj commented on drivers/sensor/tmp117/tmp117.c:
```diff
@@ -0,0 +1,189 @@
+			cfg->i2c_addr);
+}
+
+
```
One newline too much.

cfriedt reviewed (changes requested):
As you've pointed out, there is a lot of similarity between the tmp116 and the tmp117.  The only significant difference that I'm aware of is precision. 

Just googling around, and it seems that other projects have just [combined the two drivers](http://222.180.149.222:82/dsh/groundstation_rtos/commit/5b1a5243486449ee3ec0ffa14d765aec7d6b925a)

Rather than duplicating everything, it might make more sense to just combine them in your PR as well.

Otherwise, looks great!

mniestroj reviewed (commented):
Would be nice to prepare that driver for multiple instances.

puranjaymohan committed:
drivers: sensor: tmp116: Add support for SENSOR_ATTR_OFFSET.

TMP117 from TI is a sensor similar to tmp116, but it has more precision
and it also supports the offset register.
Add support for SENSOR_ATTR_OFFSET, which is used by TMP117 to the
pre-exisiting tmp116 driver.

This commit also enables multiple instances and adds TMP117 to Kconfig.

Signed-off-by: Puranjay Mohan <puranjay12@gmail.com>

puranjaymohan committed:
samples: sensor: tmp116: Add sample code for tmp117

Add sample code which demostrates the use of the offset
register offered in the TMP117.
Also add information about tmp117 in README.

Signed-off-by: Puranjay Mohan <puranjay12@gmail.com>

puranjaymohan commented:
@cfriedt I have combined the two drivers, i.e added support for tmp117 in the existing tmp116 driver, Also added some code in the sample to demonstrate tmp117.

@mniestroj I have resolved everything that you suggested.

cfriedt commented on drivers/sensor/tmp116/tmp116.c:
```diff
@@ -36,31 +36,42 @@ static int tmp116_reg_read(const struct device *dev, uint8_t reg,
  * @retval -EIO Otherwise
  */
-static inline int tmp116_device_id_check(const struct device *dev)
+static inline uint16_t tmp116_device_id_check(const struct device *dev)
```
I think it would be fine to keep this as an `int`. Generally, in Zephyr, a negative value is used for errors. Since the `int` is also capable of representing a `uint16_t`, that is covered.

puranjaymohan replied:
Sure, I will change it.

cfriedt commented on drivers/sensor/tmp116/tmp116.c:
```diff
@@ -143,8 +189,9 @@ static struct tmp116_data tmp116_data;
 };
 
-DEVICE_AND_API_INIT(tmp116, DT_INST_LABEL(0), tmp116_init,
+DEVICE_DT_INST_DEFINE(0, tmp116_init, device_pm_control_nop,
```
Looks like you'll want to modify this line so that the instance number is not hard-coded.

I would suggest creating another macro like this:
```cpp
#define DEFINE_TMP116(_num) \
    DEVICE_DT_INST_DEFINE(_num, ... \
```

 and then add the following macro underneath.

```cpp
DT_INST_FOREACH_STATUS_OKAY(DEFINE_TMP116);
```

That will allow for multiple device instances. Also ensure you remove hard-coded instance numbers elsewhere.

puranjaymohan replied:
Done, tested it on hardware.

cfriedt commented on drivers/sensor/tmp116/tmp116.h:
```diff
@@ -24,13 +25,19 @@
 #define TMP116_DEVICE_ID		0x1116
 #define TMP117_DEVICE_ID		0x0117
 
+#define DEFINE_TMP116(_num) \
```
This should go in your .c file, it should be used to initialize compile-time const items in your `device->config`.

There are a number of examples in the codebase to help clarify. E.g. 
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/i2c/i2c_emul.c#L137

For someone coming from the Linux world, the link above might shed some light on things as it highlights Zephyr's usage of DeviceTree to bake in compile-time const values into ROM and how devices are compile-time const pointers (If a device is not referenced, its section can be discarded).

The documentation could probably be improved, but here is the link to that.
https://docs.zephyrproject.org/latest/reference/drivers/index.html#c.DEVICE_DT_INST_DEFINE

puranjaymohan replied:
I have done all the changes, now I see that both device->config and device->data will be generated using the device tree by the preprocessor for each instance.

cfriedt reviewed (changes requested):
I made some comments about the device init macros. Please have a look.

gmarull commented on drivers/sensor/tmp116/tmp116.c:
```diff
@@ -112,21 +125,56 @@ static int tmp116_channel_get(const struct device *dev,
+		return -ENOTSUP;
+	}
+
+	return 0;
```
this return path is never reached

puranjaymohan replied:
corrected.

gmarull commented on drivers/sensor/tmp116/tmp116.c:
```diff
@@ -36,31 +36,44 @@ static int tmp116_reg_read(const struct device *dev, uint8_t reg,
 static inline int tmp116_device_id_check(const struct device *dev)
 {
-	uint16_t value;
+	int value;
```
use `uint16_t` here, `tmp116_reg_read` accepts `uint16_t`

puranjaymohan replied:
done.

gmarull commented on drivers/sensor/tmp116/tmp116.c:
```diff
@@ -36,31 +36,44 @@ static int tmp116_reg_read(const struct device *dev, uint8_t reg,
  * @param[in]   dev     Pointer to the device structure
  *
- * @retval 0 On success
+ * @retval device id On success
```
I think it's better to use an argument for output value and leave return code for errors only

puranjaymohan replied:
Used argument to return the ID and used return value for error code as you suggested.

gmarull commented on drivers/sensor/tmp116/tmp116.c:
```diff
@@ -36,27 +36,36 @@ static int tmp116_reg_read(const struct device *dev, uint8_t reg,
  * @brief Check the Device ID
  *
  * @param[in]   dev     Pointer to the device structure
  *
```
list of params needs to be updated

puranjaymohan replied:
done

gmarull reviewed (commented):
It would be nice if the driver can support multiple compatibles, so that it is clear that both TMP116/117 are supported. If not, maybe updating Kconfig/dt-bindings help strings would be sufficient for now.

gmarull commented on drivers/sensor/tmp116/tmp116.c:
```diff
@@ -112,39 +122,78 @@ static int tmp116_channel_get(const struct device *dev,
 	if (rc < 0) {
 		return rc;
+	} else {
+		LOG_INF("Got device ID: %x", id);
```
better use `LOG_DBG`, `else` branch not needed.

puranjaymohan replied:
done

gmarull reviewed (approved):
LGTM, thanks!

cfriedt reviewed (approved):
LGTM2!

mniestroj commented on drivers/sensor/tmp116/tmp116.c:
```diff
@@ -112,39 +122,77 @@ static int tmp116_channel_get(const struct device *dev,
+
+	switch (attr) {
+	case SENSOR_ATTR_OFFSET:
+		tmp116_device_id_check(dev, &id);
```
I think it would be better to cache device id in device data, instead of trying to fetch device id whenever SENSOR_ATTR_OFFSET is being changed.

puranjaymohan replied:
I added a variable to store the id in device data.

mniestroj reviewed (changes requested):
`Add support for TMP117.` -> this commit title is no longer valid. There was already support for TMP117. This PR/commit adds support for setting `SENSOR_ATTR_OFFSET`, multiple instances and Kconfig update.

Code changes look okay.

mniestroj commented on samples/sensor/tmp116/src/main.c:
```diff
@@ -14,13 +14,29 @@ void main(void)
+			      SENSOR_ATTR_OFFSET, &offset_value);
+	if (ret) {
+		printk("sensor_attr_set failed ret %d\n", ret);
+		return;
```
This sample will simply return early when used with `tmp116`, because sensor_attr_set will return `-EINVAL`.

puranjaymohan replied:
I have corrected this.

mniestroj commented on samples/sensor/tmp116/src/main.c:
```diff
@@ -14,13 +14,29 @@ void main(void)
 
+	/*
+	 * if an offset of 2.5 oC is to be added, set val1 = 2 and val2 = 5
+	 */
```
This is wrong. According to `struct sensor_value` documentation you should use `val1 = 2` and `val2 = 500000` in that case. This means that code implementing `SENSOR_ATTR_OFFSET` for tmp117 is also invalid.

puranjaymohan replied:
I changed the sample code and also fixed the driver to use val1 + val2^-6.

mniestroj commented on samples/sensor/tmp116/src/main.c:
```diff
@@ -14,13 +14,31 @@ void main(void)
+	 * See struct sensor_value documentation for more details.
+	 */
+	offset_value.val1 = 0;
+	offset_value.val2 = 300000;
```
It will be confusing for users to have 0.3 offset. I think 0.0 offset (as it was before) makes more sense for this sample.

puranjaymohan replied:
ohh.. sorry, I forgot to change that.. I was testing it so had it at 0.3

puranjaymohan replied:
changed it.
